cmd/zrepl/zrepl.c

```diff
@@ -0,0 +1,976 @@
+	istgt_addr.sin_port = htons(6060);
+	free(buf);
+retry:
+	rc = connect(sfd, (struct sockaddr *)&istgt_addr, sizeof (istgt_addr));
+	if ((rc == -1) && (errno == EINTR)) {
+		ZREPL_ERRLOG("Failed to connect to istgt_controller"
+		    " with err:%d\n", errno);
+		sleep(10);
+		goto retry;
+	} else {
+		printf("Connection to TGT controller successful\n");
+		ZREPL_LOG("Connection to TGT controller iss successful\n");
+	}
+
+	efd = epoll_create1(0);
+	if (efd == -1) {
+		ZREPL_ERRLOG("epoll_create() failed with errno:%d\n", errno);
+		goto exit;
+	}
+
+	event.data.fd = sfd;
+	event.events = EPOLLIN | EPOLLET | EPOLLERR |
+	    EPOLLHUP | EPOLLRDHUP;
+	rc = epoll_ctl(efd, EPOLL_CTL_ADD, sfd, &event);
+	if (rc == -1) {
+		ZREPL_ERRLOG("epoll_ctl() failed with errno:%d\n", errno);
+		goto exit;
+	}
+
+	/* Buffer where events are returned */
+	events = calloc(MAXEVENTS, sizeof (event));
+
+	/* The event loop */
+	while (1) {
+		int n, i;
+		n = epoll_wait(efd, events, MAXEVENTS, -1);
+		for (i = 0; i < n; i++) {
+			if ((events[i].events & EPOLLERR) ||
+			    (events[i].events & EPOLLHUP) ||
+			    (events[i].events & EPOLLRDHUP)) {
+				/*
+				 * Error has occurred on this socket
+				 * close it and open a new socket after
+				 * 5 sec of sleep.
+				 */
+				ZREPL_ERRLOG("epoll err() :%d\n", errno);
+				close(events[i].data.fd);
+				printf("Retrying ....\n");
+				sleep(5);
+				sfd = create_and_bind(mgmt_port);
+				if (sfd == -1) {
+					goto exit;
+				}
+
+				rc = make_socket_non_blocking(sfd);
+				if (rc == -1) {
+					goto exit;
+				}
+retry1:
+				rc = connect(sfd,
+				    (struct sockaddr *)&istgt_addr,
+				    sizeof (istgt_addr));
+				if ((rc == -1) && (errno == EINTR)) {
+					ZREPL_ERRLOG("Failed to connect to"
+					    " istgt_controller with err:%d\n",
+					    errno);
+					sleep(2);
+					goto retry1;
+				}
+
+				event.data.fd = sfd;
+				event.events = EPOLLIN | EPOLLET |
+				    EPOLLERR | EPOLLHUP | EPOLLRDHUP;
+				rc = epoll_ctl(efd, EPOLL_CTL_ADD, sfd, &event);
+				if (rc == -1) {
+					ZREPL_ERRLOG("epoll_ctl() failed with"
+					    " errno:%d\n", errno);
+					goto exit;
+				}
+				continue;
+			}
+
+			bzero(&hdr, sizeof (hdr));
+			count = read(events[i].data.fd, (char *)&hdr,
+			    sizeof (hdr));
+			if (count == -1) {
+				ZREPL_ERRLOG("Read from socket failed"
+				    " with err:%d\n", errno);
+				goto exit;
+			}
+
+			if (hdr.opcode == ZVOL_OPCODE_HANDSHAKE) {
+				name = kmem_alloc(
+				    hdr.len * sizeof (char), KM_SLEEP);
+				count = read(events[i].data.fd, name,
+				    sizeof (char) * hdr.len);
+				if (count == -1) {
+					ZREPL_ERRLOG("Read from socket failed"
+					    " with err:%d\n", errno);
+					goto exit;
+				}
+
+				rc = uzfs_zvol_mgmt_do_handshake(&hdr,
+				    events[i].data.fd, name);
+				free(name);
+				if (rc == -1) {
+					ZREPL_ERRLOG("handshake failed with"
+					    " errno:%d\n", errno);
+					goto exit;
+				}
+			}
```

Comment:
IIRC the file descriptor must be inserted to epoll set again if you want to continue receiving events on it.

Reply:
Jan, I got confused, are we not putting fd back to event channel at line 593 ? Nonetheless I have changed this code completely (because of previous comment, I have made this fd blocking). Request you to have a look again to see if we missed something.